Add basic logging system

mahdikarami8484 opened the pull request:
A basic logging system has been implemented with three log levels: Info, Warning, and Error.

To use the logger, simply include the `core/debug/debug.h` header and call:
    Debug::Info("Your message");

Macro-based logging is also available, which automatically includes file name and line number:
    LOG_INFO("Message");

Automatic disabling of logs in Release builds has been added using `NDEBUG`, although this part still needs improvement for better extensibility and maintainability.

Log level filtering is supported via:
    Debug::SetLevel(LogLevel::Warning);

---

TODOs:
- Add support for saving logs to a file.

mahdikarami8484 committed:
Add basic logging system

A basic logging system has been implemented with three log levels: Info, Warning, and Error.

To use the logger, simply include the `core/debug/debug.h` header and call:
    Debug::Info("Your message");

Macro-based logging is also available, which automatically includes file name and line number:
    LOG_INFO("Message");

Automatic disabling of logs in Release builds has been added using `NDEBUG`, although this part still needs improvement for better extensibility and maintainability.

Log level filtering is supported via:
    Debug::SetLevel(LogLevel::Warning);

---

TODOs:
- Add support for saving logs to a file.

mahdikarami8484 committed:
add sublime config file

mahdikarami8484 committed:
add a definition for debug

mahdikarami8484 committed:
Rename CmakeLists.txt to CMakeLists.txt

mahdikarami8484 committed:
Rename CmakeLists.txt to CMakeLists.txt

mahdikarami8484 committed:
Rename CmakeLists.txt to CMakeLists.txt

mahdikarami8484 committed:
fix the error

MandelbrotInferno commented on src/core/debug/logger.cpp:
There are 3 major issues that I have with this design and 1 minor one which doesn't matter at the moment but will become important later on.

* 3 major issues are the following:
1)  std::cout/std::cerr are slow unless are somehow optimized. I do not know the performance of forming a message using std::ostringstream so that def also needs to be benchmarked.
2) The use of std::string here is ill advised since that will most likely cause a lot of heap allocations. That alone , depending on where it is used, could slow down the app considerably. Since our application is a real time one, we need to make sure there aren't too much performance issues when debugging using our logger.
3) The user cannot do something like the following: 

```C++
Debug::Warning("Function CreateTexture() failed: %s", Error);
```

Instead they need to format their texts and then pass it to the const std::string& message. That is not easy to use and def not user friendly. We need to take care of the formatting in our logger as well.

* 1 Minor issue:
1) It is not thread safe to use this logger at the moment. In the near future we might have to make it thread safe by using synchronization primitives of standard library eg. std::mutex etc.


I suggest you add the Tracy profiler to the project so we could benchmark the issues I raised above.

mahdikarami8484 committed:
Add Tracy v0.12.2 as submodule

mahdikarami8484 committed:
Merge branch 'setup/build-system' into external/tracy

mahdikarami8484 committed:
Apply requested changes

Improve logging performance and add log formatting support

- Logging and console output moved to a separate thread to prevent delays in the main program
- Added support for formatted log messages
- Significantly improved performance of logging functions

Usage example can be found in PXEngine.cpp

mahdikarami8484 committed:
off loggin in release mode

mahdikarami8484 committed:
update for tracy

mahdikarami8484 committed:
Merge branch 'setup/build-system' into core/debug/logging
